File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B04_dbscan/python/B04_dbscan.py
"""
B4. DBSCAN
Tema B04_dbscan — Introducción al Machine Learning (IIND3118)

CÓMO USARLO
  1. Cambia el bloque marcado con ✏️ abajo por tu dataset.
  2. Corre el archivo:  python B04_dbscan.py
  3. Las gráficas se guardan en la carpeta figuras/ de este tema.

La explicación del método está en el README.md de esta misma carpeta.
"""

import os, sys, warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")          # no abre ventanas: funciona en cualquier entorno
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, KFold, cross_val_score, GridSearchCV
from sklearn.preprocessing import scale
from sklearn.linear_model import (LinearRegression, Ridge, Lasso, ElasticNet,
                                  RidgeCV, LassoCV, ElasticNetCV, LogisticRegression)
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.metrics import (mean_squared_error, mean_absolute_error, r2_score,
                             accuracy_score, confusion_matrix, classification_report,
                             roc_curve, roc_auc_score)
from sklearn.decomposition import PCA, KernelPCA
from sklearn.cluster import KMeans, DBSCAN
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor
from sklearn import manifold

# ── Guardado automático de figuras ──────────────────────────────────────
AQUI = os.path.dirname(os.path.abspath(__file__))
FIGS = os.path.join(AQUI, "..", "figuras")
os.makedirs(FIGS, exist_ok=True)
_n = [0]
def _save():
    _n[0] += 1
    ruta = os.path.join(FIGS, f"{_n[0]:02d}_figura.png")
    plt.savefig(ruta, dpi=110, bbox_inches="tight"); plt.close()
    print(f"   [figura guardada: figuras/{os.path.basename(ruta)}]")
plt.show = _save                          # cada plt.show() guarda en vez de abrir

RNG = np.random.default_rng(0)

# ╔═════════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                             ║
# ╚═════════════════════════════════════════════════════════════════════════╝
DATOS  = os.path.join(AQUI, "..", "..", "..", "data", "ejemplo_regresion.csv")
TARGET = "target"          # nombre EXACTO de la columna respuesta. None si no hay.
# ╚═════════════════════════════════════════════════════════════════════════╝

df = pd.read_csv(DATOS).dropna()

if TARGET is not None:
    y_serie = df[TARGET]; X_df = df.drop(columns=[TARGET])
else:
    y_serie = None; X_df = df.copy()

X_df = pd.get_dummies(X_df, drop_first=True).astype(float)
nombres_X = list(X_df.columns)
X  = X_df.values
Xs = scale(X)

if y_serie is not None:
    y = pd.factorize(y_serie)[0] if y_serie.dtype == object else y_serie.values
    ES_CLASIFICACION = (len(np.unique(y)) <= 10) and np.all(y == np.round(y))
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30, random_state=0)
    Xs_train, Xs_test, _, _ = train_test_split(Xs, y, test_size=.30, random_state=0)
else:
    y = None; ES_CLASIFICACION = False

def ModeloLineal(**kw):
    return (LogisticRegression(max_iter=5000, **kw) if ES_CLASIFICACION
            else LinearRegression(**kw))

col = y if y is not None else "steelblue"
Z   = PCA().fit_transform(Xs)        # proyección PCA, se usa para graficar
Z2  = Z[:, :2]

print("="*74)
print("B4. DBSCAN")
print("="*74)
print(f"datos: {X.shape[0]} obs. x {X.shape[1]} predictores")
if y is not None:
    print(f"y ('{TARGET}'): " + ("CLASIFICACIÓN" if ES_CLASIFICACION else "REGRESIÓN"))
print()

# ════════════════════════════════════════════════════════════════════════
#  CÓDIGO DEL MÉTODO
# ════════════════════════════════════════════════════════════════════════

# ═══ B4. DBSCAN ═══
from sklearn.neighbors import NearestNeighbors

# ── Paso 1: curva k-dist para elegir eps ───────────────────────────────
k = min(5, len(Xs)-1)
d_k = np.sort(NearestNeighbors(n_neighbors=k).fit(Xs).kneighbors(Xs)[0][:,1:].reshape(-1))

plt.plot(d_k); plt.xlabel("puntos ordenados"); plt.ylabel(f"distancia al vecino k={k}")
plt.title("Curva k-dist: busca el CODO para elegir eps"); plt.show()
print(f"Sugerencia: eps entre {np.percentile(d_k,80):.2f} (percentil 80) y "
      f"{np.percentile(d_k,95):.2f} (percentil 95)")

# ── Paso 2: probar varios eps ──────────────────────────────────────────
print("\nSensibilidad a eps:")
for eps in np.round(np.percentile(d_k, [50,70,80,90,95]), 2):
    lab = DBSCAN(eps=eps, min_samples=k+1).fit_predict(Xs)
    n_cl = len(set(lab)) - (1 if -1 in lab else 0)
    print(f"  eps={eps:5.2f} -> {n_cl} clusters, {int((lab==-1).sum())} puntos de ruido")

# ── Paso 3: modelo final ───────────────────────────────────────────────
EPS = float(np.percentile(d_k, 90))     # ⬅ ajusta según la curva y la tabla de arriba
db = DBSCAN(eps=EPS, min_samples=k+1).fit(Xs)

n_cl = len(set(db.labels_)) - (1 if -1 in db.labels_ else 0)
print(f"\nFINAL: eps={EPS:.2f} -> {n_cl} clusters, {int((db.labels_==-1).sum())} ruido")

mask = db.labels_ == -1
plt.scatter(Z2[~mask,0], Z2[~mask,1], c=db.labels_[~mask], cmap="viridis", alpha=.7)
plt.scatter(Z2[mask,0], Z2[mask,1], c="red", marker="x", s=60, label="ruido (-1)")
plt.xlabel("PC1"); plt.ylabel("PC2"); plt.title(f"DBSCAN (eps={EPS:.2f})")
plt.legend(); plt.show()

print("\nListo. Revisa la carpeta figuras/")

File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B04_dbscan/R/B04_dbscan.R
# ############################################################################
# B4. DBSCAN
# Tema B04_dbscan — Introducción al Machine Learning (IIND3118)
#
# CÓMO USARLO
#   1. Cambia el bloque marcado con ✏️ por tu dataset.
#   2. Corre:  Rscript B04_dbscan.R     (o ábrelo en RStudio)
#   3. Las gráficas se guardan en la carpeta figuras/ de este tema.
#
# La explicación del método está en el README.md de esta misma carpeta.
# ############################################################################

for (p in c("MASS","cluster","dbscan","glmnet","leaps","caret","class","fpc",
            "clusterCrit","isotree","kernlab","Rtsne","psych","car","vegan")) {
  if (!requireNamespace(p, quietly=TRUE))
    try(install.packages(p, repos="https://cloud.r-project.org"), silent=TRUE)
}
suppressMessages({ library(MASS); library(cluster) })

set.seed(0)

# Las gráficas se guardan en figuras/ en vez de abrirse en pantalla
dir.create("../figuras", showWarnings=FALSE)
.n <- 0
guardar <- function() {
  .n <<- .n + 1
  dev.copy(png, sprintf("../figuras/%02d_figura.png", .n), width=900, height=650)
  invisible(dev.off())
}

# ╔═══════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                           ║
# ╚═══════════════════════════════════════════════════════════════════════╝
DATOS  <- "../../../data/ejemplo_regresion.csv"
TARGET <- "target"        # nombre EXACTO de la columna respuesta. NULL si no hay.
# ╚═══════════════════════════════════════════════════════════════════════╝

df <- na.omit(read.csv(DATOS))

if (!is.null(TARGET)) {
  y_raw <- df[[TARGET]]
  X_df  <- df[, setdiff(names(df), TARGET), drop=FALSE]
} else { y_raw <- NULL; X_df <- df }

X_df <- as.data.frame(model.matrix(~ . - 1, data=X_df))
nombres_X <- names(X_df)
X  <- as.matrix(X_df)
Xs <- scale(X)

if (!is.null(y_raw)) {
  ES_CLASIFICACION <- is.factor(y_raw) || is.character(y_raw) || length(unique(y_raw)) <= 10
  y <- if (ES_CLASIFICACION) as.factor(y_raw) else as.numeric(y_raw)
  idx <- sample(seq_len(nrow(X)), floor(0.70*nrow(X)))
  Xtrain <- X[idx,]; Xtest <- X[-idx,]; ytrain <- y[idx]; ytest <- y[-idx]
  Xs_train <- Xs[idx,]; Xs_test <- Xs[-idx,]
} else { y <- NULL; ES_CLASIFICACION <- FALSE }

pca  <- prcomp(X, center=TRUE, scale.=TRUE)
Z2   <- pca$x[, 1:2]
col_pt <- if (!is.null(y)) as.integer(as.factor(y)) else "steelblue"

cat(strrep("=",74), "\n")
cat("B4. DBSCAN\n")
cat(strrep("=",74), "\n")
cat(sprintf("datos: %d obs. x %d predictores\n\n", nrow(X), ncol(X)))

# ############################################################################
#  CÓDIGO DEL MÉTODO
# ############################################################################

# ============================================================================
# B4. DBSCAN
# ----------------------------------------------------------------------------
# QUÉ HACE: busca ZONAS DENSAS. Un cluster es un montón de puntos pegados.
# No exige que todos estén cerca del centro, solo que haya una CADENA de
# vecinos densos. Por eso encuentra FORMAS RARAS (espirales, anillos).
# Además marca como RUIDO los puntos que no caen en ninguna zona densa.
# Es el ÚNICO método de clustering que dice "este punto no va en ningún lado".
#
# LOS 3 TIPOS DE PUNTO
#   core   : tiene al menos minPts vecinos dentro del radio eps
#   border : no es core, pero está dentro del radio de un core
#   noise  : ninguna de las dos -> etiqueta 0 EN R  (⚠️ en Python es -1)
#
# CÓMO ELEGIR eps: LA CURVA k-dist
#   1. distancia de cada punto a su k-ésimo vecino
#   2. ordenar y graficar  ->  kNNdistplot()
#   3. el CODO de la curva es un buen eps
#
# PARÁMETROS
#   eps    : ⚠️ EL CRÍTICO. Radio de la vecindad. Se lee del codo.
#   minPts : regla práctica >= p+1, o 2p si hay ruido
#
# DIAGNÓSTICO: casi todo es ruido -> eps muy chico.
#              un solo cluster gigante -> eps muy grande.
#
# ✅ Formas arbitrarias · no fija K · detecta outliers
# ❌ Calibración difícil y MUY sensible ·
#    FALLA si los clusters tienen densidades distintas (usa hdbscan) ·
#    se degrada en dimensión alta
# ============================================================================

library(dbscan)
k_vec <- min(5, nrow(Xs)-1)

# Paso 1: la curva k-dist
kNNdistplot(Xs, k=k_vec)
title("Curva k-dist: busca el CODO para elegir eps")
d_k <- sort(as.vector(kNNdist(Xs, k=k_vec)))
abline(h=quantile(d_k, .90), col="red", lty=2)
cat(sprintf("Sugerencia: eps entre %.2f (p80) y %.2f (p95)\n",
            quantile(d_k,.80), quantile(d_k,.95)))

# Paso 2: probar varios eps
cat("\nSensibilidad a eps:\n")
for (eps in round(quantile(d_k, c(.5,.7,.8,.9,.95)), 2)) {
  lab <- dbscan(Xs, eps=eps, minPts=k_vec+1)$cluster
  cat(sprintf("  eps=%5.2f -> %d clusters, %d puntos de ruido\n",
              eps, length(setdiff(unique(lab),0)), sum(lab==0)))
}

# Paso 3: modelo final
EPS <- as.numeric(quantile(d_k, .90))     # ⬅ ajusta según la curva
db  <- dbscan(Xs, eps=EPS, minPts=k_vec+1)
cat(sprintf("\nFINAL: eps=%.2f -> %d clusters, %d ruido\n",
            EPS, length(setdiff(unique(db$cluster),0)), sum(db$cluster==0)))

plot(Z2, col=ifelse(db$cluster==0, "red", db$cluster+1),
     pch=ifelse(db$cluster==0, 4, 16),
     xlab="PC1", ylab="PC2", main=sprintf("DBSCAN (eps=%.2f)", EPS))
legend("topright", "ruido (0)", col="red", pch=4)

# Si las densidades son muy distintas: hdbscan(Xs, minPts=5)

cat("\nListo. Revisa la carpeta figuras/\n")
